README.md
# rollup-plugin-url

Inline import files as data-URIs, or copy them to output

## Install

```sh
npm i --save-dev rollup-plugin-url
```

## Usage

```js
import {rollup} from "rollup"
import url from "rollup-plugin-url"

const writeoptions = {dest: "output/output.js"}
const plugin = url({
  limit: 10 * 1024, // inline files < 10k, copy files > 10k
  include: ["**/*.svg"], // defaults to .svg, .png, .jpg and .gif files
  emitFiles: true // defaults to true
})

rollup({
  entry: "main.js",
  plugins: [plugin],
})
.then(bundle => bundle.write(writeoptions))
```

## Options

### limit

Optional. Type: `number`.

This is the file size limit to inline files. If files exceed this limit, they
will be copied instead to the destination folder and the hashed filename will
be given instead. If value set to `0` all files will be copied.

Defaults to 14336 (14kb).

### include / exclude

Optional. Type: a minimatch pattern, or array of minimatch patterns

These patterns determine which files are inlined. Defaults to .svg, .png, .jpg
and .gif files.

### publicPath

Optional. Type: `string`

The `publicPath` will be added in front of file names when they are not inlined
but copied.

### emitFiles

Optional. Type: `boolean`

The `emitFiles` option is used to run the plugin as you normally would but prevents any files being emitted. This is useful for when you are using rollup to emit both a client side and server side bundle.

### fileName

Optional. Type: `string`

When `emitFiles` is `true`, the `fileName` option can be used to rename the emitted files. It accepts the following string replacements:

- `[hash]` - The hash value of the file's contents
- `[name]` - The name of the imported file, without it's file extension
- `[extname]` - The extension of the imported file, including the leading `.`
- `[dirname]` - The parent directory name of the imported file, including trailing `/`

Defaults to: `"[hash][extname]"`

### sourceDir

Optional. Type: `string`

When using the `[dirname]` replacement in `fileName`, uses this directory as the source directory to create the file path from rather than the parent directory of the imported file. For example:

*src/path/to/file.js*
```js
import png from "./image.png";
```
*rollup.config.js*
```js
url({
  fileName: "[dirname][hash][extname]",
  sourceDir: path.join(__dirname, "src")
})
```
Emitted File: `path/to/image.png`

# License

LGPL-3.0


CHANGELOG.md
# Changelog

2.2.4 / 2019-10-08
=================
* fallback to `mkdirp` to keep working on node <= 8

2.2.3 / 2019-10-08
=================
* remove `mkpath` from dependencies ([#24](https://github.com/rollup/rollup-plugin-url/pull/24))
* update dev dependencies

2.2.2 / 2019-06-13
=================
* dependencies update

2.2.1 / 2019-04-10
=================
* Fix `dirname` substitution on Windows ([#21](https://github.com/rollup/rollup-plugin-url/pull/21))

2.2.0 / 2019-01-29
=================
* add `destDir` option ([#19](https://github.com/rollup/rollup-plugin-url/pull/19))
* update dependencies

2.1.0 / 2018-12-02
==================
* add `fileName` option ([#17](https://github.com/rollup/rollup-plugin-url/pull/17))

2.0.1 / 2018-10-09
==================
* ensure destination folder exist while `generateBundle` hook performs

2.0.0 / 2018-10-01
==================
* **Breaking:** version 2.0.0 requires rollup@0.60 and higher – deprecated `onwrite` hook replaced with new `generateBundle` hook, so plugin will not work with earlier versions of rollup.  
Use version 1.4 with rollup<0.60

1.4.0 / 2018-04-17
==================
 * add support for `output.dir` option
 * update dependencies

1.3.0 / 2017-09-17
==================
 * internal update: it builds now with rollup@0.50 

1.2.0 / 2017-06-09
==================

 * add ability to prevent emitting any files with the `emitFiles=false` option.

1.1.0 / 2017-04-12
==================

 * set default limit to 14kb

1.0.0 / 2017-04-10
==================

 * Migrate to newer rollup API (#5).
 * Minimal `rollup` version is `0.32.4`
 * Braking: `write` method was removed

0.1.2 / 2016-08-30
==================

  * Add public path option (#1)

0.1.1 / 2016-02-08
==================

  * drop charset, its not needed

0.1.0 / 2016-02-08
==================

  * Initial release


test/index.js
import assert from "assert"
import fs from "fs"
import path from "path"
import rimraf from "rimraf"
import {rollup} from "rollup"
import url from "../"

const dest = "output/output.js"
const dir = "output/splitting"

process.chdir(__dirname)

const svghash = "98ea1a8cc8cd9baf.svg"
const pnghash = "6b71fbe07b498a82.png"
const pngname = "png.png"

const asserts = {
  svgInline: `var svg = "data:image/svg+xml,%3Csvg%3E%3Cpath%20d%3D%22%22%2F%3E%3C%2Fsvg%3E";\nexport default svg;`,
  svgExport: `var svg = "${svghash}";\nexport default svg;`,
  pngInline: `var png = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAAAAAA6fptVAAAACklEQVR4nGP6DwABBQECz6AuzQAAAABJRU5ErkJggg==";\nexport default png;`,
  pngExport: `var png = "${pnghash}";\nexport default png;`
}

describe("rollup-plugin-url", () => {
  after(() => promise(rimraf, "output/"))

  it("should inline text files", () =>
    run("./fixtures/svg.js", 10 * 1024)
      .then(
        () => Promise.all([
          assertOutput(asserts.svgInline),
          assertExists(`output/${svghash}`, false),
        ])
      )
  )

  it("should not copy files when limit is 0 and emitFiles is off", () =>
    run("./fixtures/svg.js", { limit: 0, publicPath: "", emitFiles: false })
      .then(
        () => Promise.all([
          assertOutput(asserts.svgExport),
          assertExists(`output/${svghash}`, false),
        ])
      )
  )

  it("should copy files when limit is 0", () =>
    run("./fixtures/svg.js", { limit: 0 })
      .then(
        () => Promise.all([
          assertOutput(asserts.svgExport),
          assertExists(`output/${svghash}`),
        ])
      )
  )

  it("should inline binary files", () =>
    run("./fixtures/png.js", { limit: 10 * 1024 })
      .then(
        () => Promise.all([
          assertOutput(asserts.pngInline),
          assertExists(`output/${pnghash}`, false),
        ])
      )
  )

  it("should copy large text files", () =>
    run("./fixtures/svg.js", { limit: 10 })
      .then(
        () => Promise.all([
          assertOutput(asserts.svgExport),
          assertExists(`output/${svghash}`),
        ])
      )
  )

  it("should copy large binary files", () =>
    run("./fixtures/png.js", { limit: 10 })
      .then(
        () => Promise.all([
          assertOutput(asserts.pngExport),
          assertExists(`output/${pnghash}`),
        ])
      )
  )

  it("should use publicPath", () =>
    run("./fixtures/png.js", { limit: 10, publicPath: "/foo/bar/" })
      .then(
        () => Promise.all([
          assertOutput(`var png = "/foo/bar/${pnghash}";\nexport default png;`),
        ])
      )
  )

  it("should create a nested directory for the output, if required", () =>
    run("./fixtures/png.js", { limit: 10, fileName: "subdirectory/[hash][extname]" })
      .then(
        () => Promise.all([
          assertExists(`output/subdirectory/${pnghash}`)
        ])
      )
  )

  it("should create a file with the name and extension of the file", () =>
    run("./fixtures/png.js", { limit: 10, fileName: "[name][extname]" })
      .then(
        () => Promise.all([
          assertExists(`output/${pngname}`)
        ])
      )
  )

  it("should create a file with the name, hash and extension of the file", () =>
    run("./fixtures/png.js", { limit: 10, fileName: "[name]-[hash][extname]" })
      .then(
        () => Promise.all([
          assertExists(`output/png-${pnghash}`)
        ])
      )
  )

  it("should prefix the file with the parent directory of the source file", () =>
    run("./fixtures/png.js", { limit: 10, fileName: "[dirname][hash][extname]" })
      .then(
        () => Promise.all([
          assertExists(`output/fixtures/${pnghash}`)
        ])
      )
  )

  it("should prefix the file with the parent directory of the source file, relative to the sourceDir option", () =>
    run("./fixtures/png.js", { limit: 10, fileName: "[dirname][hash][extname]", sourceDir: path.join(__dirname, "../") })
      .then(
        () => Promise.all([
          assertExists(`output/test/fixtures/${pnghash}`)
        ])
      )
  )

  it("should copy the file according to destDir option", () =>
    run("./fixtures/png.js", { limit: 10, fileName: "[dirname][hash][extname]", destDir: path.join(__dirname, "output/dest") })
      .then(
        () => Promise.all([
          assertExists(`output/dest/fixtures/${pnghash}`)
        ])
      )
  )

  it("should create multiple modules and inline files", () => {
    return run(["./fixtures/svg.js", "./fixtures/png.js"], {}, true)
      .then(
        () => Promise.all([
          assertOutput(asserts.pngInline, `${dir}/png.js`),
          assertOutput(asserts.svgInline, `${dir}/svg.js`)
        ])
      )
  })

  it("should create multiple modules and copy files", () => {
    return run(["./fixtures/svg.js", "./fixtures/png.js"], { limit: 0 }, true)
      .then(
        () => Promise.all([
          assertOutput(asserts.pngExport, `${dir}/png.js`),
          assertOutput(asserts.svgExport, `${dir}/svg.js`),
          assertExists(`${dir}/${svghash}`),
          assertExists(`${dir}/${pnghash}`)
        ])
      )
  })
})

function promise(fn, ...args) {
  return new Promise((resolve, reject) =>
    fn(...args, (err, res) =>
      err ? reject(err) : resolve(res)))
}

const DEFAULT_OPTIONS = {
  publicPath: "",
  emitFiles: true
}

function run(input, options, experimentalCodeSplitting = false) {
  const writeOptions = {
    format: 'es'
  }

  if (!experimentalCodeSplitting) {
    Object.assign(writeOptions, {
      file: dest
    })
  } else {
    Object.assign(writeOptions, {
      dir
    })
  }

  return rollup({
    input,
    plugins: [url(Object.assign({}, DEFAULT_OPTIONS, options))],
    experimentalCodeSplitting
  }).then(bundle => bundle.write(writeOptions))
}

function assertOutput(content, file = dest) {
  return promise(fs.readFile, file, "utf-8")
    .then(fileContent => assert.equal(fileContent.replace(/[\W]/gi, ''), content.replace(/[\W]/gi, '')))
}

function assertExists(name, shouldExist = true) {
  return promise(fs.stat, name)
    .then(() => true, () => false)
    .then(exists => assert.ok(exists === shouldExist))
}


src/index.js
import {createFilter} from "rollup-pluginutils"
import mkdirp from 'mkdirp'
import mime from "mime"
import crypto from "crypto"
import path from "path"
import fs from "fs"

const defaultInclude = [
  "**/*.svg",
  "**/*.png",
  "**/*.jpg",
  "**/*.gif",
]

export default function url(options = {}) {
  const {
    limit = 14 * 1024,
    include = defaultInclude,
    exclude,
    publicPath = "",
    emitFiles = true,
    fileName = "[hash][extname]"
  } = options
  const filter = createFilter(include, exclude)

  const copies = Object.create(null)

  return {
    load(id) {
      if (!filter(id)) {
        return null
      }
      return Promise.all([
        promise(fs.stat, id),
        promise(fs.readFile, id),
      ]).then(([stats, buffer]) => {
        let data
        if ((limit && stats.size > limit) || limit === 0) {
          const hash = crypto.createHash("sha1")
            .update(buffer)
            .digest("hex")
            .substr(0, 16)
          const ext = path.extname(id)
          const name = path.basename(id, ext)
          // Determine the directory name of the file based
          // on either the relative path provided in options,
          // or the parent directory
          const relativeDir = options.sourceDir
            ? path.relative(options.sourceDir, path.dirname(id))
            : path.dirname(id).split(path.sep).pop()

          // Generate the output file name based on some string
          // replacement parameters
          const outputFileName = fileName
            .replace(/\[hash\]/g, hash)
            .replace(/\[extname\]/g, ext)
            .replace(/\[dirname\]/g, `${relativeDir}/`)
            .replace(/\[name\]/g, name)
          data = `${publicPath}${outputFileName}`
          copies[id] = outputFileName
        } else {
          const mimetype = mime.getType(id)
          const isSVG = mimetype === "image/svg+xml"
          data = isSVG
            ? encodeSVG(buffer)
            : buffer.toString("base64")
          const encoding = isSVG ? "" : ";base64"
          data = `data:${mimetype}${encoding},${data}`
        }
        return `export default "${data}"`
      })
    },
    generateBundle: async function write(outputOptions) {
      // Allow skipping saving files for server side builds.
      if (!emitFiles) return

      const base = options.destDir || outputOptions.dir || path.dirname(outputOptions.file)

      await promise(mkdirp, base)

      return Promise.all(Object.keys(copies).map(async name => {
        const output = copies[name]
        // Create a nested directory if the fileName pattern contains
        // a directory structure
        const outputDirectory = path.join(base, path.dirname(output))
        await promise(mkpath, outputDirectory)
        return copy(name, path.join(base, output))
      }))
    }
  }
}

function promise(fn, ...args) {
  return new Promise((resolve, reject) =>
                     fn(...args, (err, res) =>
                        err ? reject(err) : resolve(res)))
}

function copy(src, dest) {
  return new Promise((resolve, reject) => {
    const read = fs.createReadStream(src)
    read.on("error", reject)
    const write = fs.createWriteStream(dest)
    write.on("error", reject)
    write.on("finish", resolve)
    read.pipe(write)
  })
}

// https://github.com/filamentgroup/directory-encoder/blob/master/lib/svg-uri-encoder.js
function encodeSVG(buffer) {
  return encodeURIComponent(buffer.toString("utf-8")
    // strip newlines and tabs
    .replace(/[\n\r]/gmi, "")
    .replace(/\t/gmi, " ")
    // strip comments
    .replace(/<!\-\-(.*(?=\-\->))\-\->/gmi, "")
    // replace
    .replace(/'/gmi, "\\i"))
    // encode brackets
    .replace(/\(/g, "%28").replace(/\)/g, "%29")
}

// use fs.mkdir to instead of mkpath package, see https://github.com/jrajav/mkpath/issues/6
function mkpath(path, err) {
  return fs.mkdir(path, { recursive: true }, err);
}
